src/Microsoft.ML.Core/Utilities/Utils.cs

```diff
@@ -751,6 +763,41 @@ public static void BuildSubsetMaps(int lim, Func<int, bool> pred, out int[] map,
             map = mapList.ToArray();
         }
 
+        /// <summary>
+        /// Given a predicate, over a range of values defined by a limit calculate
+        /// first the values for which that predicate was true, and second an inverse
+        /// map.
+        /// </summary>
+        /// <param name="lim">Indicates the exclusive upper bound on the tested values</param>
+        /// <param name="colsNeeded">The set of columns the calling component operates on.</param>
+        /// <param name="map">An ascending array of values from 0 inclusive
+        /// to <paramref name="lim"/> exclusive, holding all values for which
+        /// <paramref name="colsNeeded"/> are present.
+        /// (The respective index appears in the <paramref name="colsNeeded"/> collection).</param>
+        /// <param name="invMap">Forms an inverse mapping of <paramref name="map"/>,
+        /// so that <c><paramref name="invMap"/>[<paramref name="map"/>[i]] == i</c>,
+        /// and for other entries not appearing in <paramref name="map"/>,
+        /// <c><paramref name="invMap"/>[i] == -1</c></param>
+        public static void BuildSubsetMaps(int lim, IEnumerable<Schema.Column> colsNeeded, out int[] map, out int[] invMap)
+        {
+            Contracts.CheckParam(lim >= 0, nameof(lim));
+
+            // REVIEW: Better names?
+            List<int> mapList = new List<int>();
+            invMap = new int[lim];
+            for (int c = 0; c < lim; ++c)
+            {
+                if (colsNeeded == null || !colsNeeded.Any(x => x.Index == c))
+                {
+                    invMap[c] = -1;
+                    continue;
+                }
+                invMap[c] = mapList.Count;
+                mapList.Add(c);
+            }
```

Comment:
Can you have foreach loop over colsNeeded instead of this. No ones guarantee you cached in memory IEnumerable, it can be anything including some over network connection collection and with trigger to start nuclear war if you access it multiple times within short period of time. You don't want to start nuclear war, right?

Reply:
Hummmm, idk if i want to start a nuclear war. I'll go literally sleep on it :)

---
In reply to: [246979630](https://github.com/dotnet/machinelearning/pull/1975#discussion_r246979630) [](ancestors = 246979630)